### PR TITLE
ddtrace/tracer: account for resource, tags, and target_span when applying sampling rules

### PR DESCRIPTION
**Pull request by dianashevchenko**
### What does this PR do?
This PR changes the sampling rule matching logic. Tests are updated accordingly.
Additionally, updated the MarshalJSON method of the SamplingRule. 

What to remember when reviewing 
- three fields are added to the sampling rule - tags, resource, target_span
- target_span is set only on trace rule
- the only (at least intended) change in regexes is I added `"target_span":"any",` where it was applicable

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.

For Datadog employees:

- [ ] If this PR touches code that handles credentials of any kind, such as Datadog API keys, I've requested a review from `@DataDog/security-design-and-guidance`.
- [ ] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: ddtrace/tracer/rules_sampler.go [hunk 1]
```diff
@@ -87,6 +87,9 @@ type SamplingRule struct {
 
 // match returns true when the span's details match all the expected values in the rule.
 func (sr *SamplingRule) match(s *span) bool {
+	if sr.TargetRoot && s.root().SpanID != s.SpanID {
```
Comment (knusbaum) [inline ddtrace/tracer/rules_sampler.go:90]: How likely is `s.root()` to return `nil`, or for `s` to be `nil`?

File: ddtrace/tracer/rules_sampler.go [hunk 2]
```diff
@@ -97,6 +100,17 @@ func (sr *SamplingRule) match(s *span) bool {
...
+	if sr.Resource != nil && !sr.Resource.MatchString(s.Resource) {
+		return false
+	}
+	if sr.Tags != nil {
+		for k, regex := range sr.Tags {
+			v, ok := s.Meta[k]
+			if !ok || !regex.MatchString(v) {
+				return false
+			}
+		}
```
Comment (knusbaum) [inline ddtrace/tracer/rules_sampler.go:112]: It looks like this should match both `Meta` and `Metrics`.

It's not specified in the RFC how numbers should be represented. Maybe talk with the authors about how this should work.

**Review by knusbaum — COMMENTED** [review]
A couple comments.

I think there's some stuff I'm not familiar with in `shevchenko/tag-resource-sampling` such as `TagsResourceRule` that it would be helpful to have a PR up for as well.

File: ddtrace/tracer/rules_sampler.go (hunk 1 above)
Comment (dianashevchenko) [reply ddtrace/tracer/rules_sampler.go:90]: Span can be newly created with all the necessary / fields / context filled out, but never nil. Subsequently, root wouldn't be nil. The match function is called only within sampling functionality on a trace / chunk, and thus span is never empty.

**Pull request merged**